hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/PutAllOperation.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.multimap.impl.operations;
+
+import com.hazelcast.core.EntryEventType;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.map.impl.MapEntries;
+import com.hazelcast.multimap.impl.MultiMapContainer;
+import com.hazelcast.multimap.impl.MultiMapDataSerializerHook;
+import com.hazelcast.multimap.impl.MultiMapRecord;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.spi.impl.operationservice.BackupAwareOperation;
+import com.hazelcast.spi.impl.operationservice.MutatingOperation;
+import com.hazelcast.spi.impl.operationservice.Operation;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Iterator;
+
+public class PutAllOperation extends AbstractMultiMapOperation implements MutatingOperation, BackupAwareOperation {
+    private transient int currentIndex;
+    private MapEntries mapEntries;
+
+    public PutAllOperation() {
+        super();
+    }
+
+    public PutAllOperation(String name, MapEntries mapEntries) {
+        super(name);
+        this.mapEntries = mapEntries;
+        //NB: general structure copied from c.hz.map.impl.operation.PutAllOperation
+    }
+
+    @Override
+    public final void run() {
+        int size = mapEntries != null ? mapEntries.size() : 0;
+        while (currentIndex < size) {
+            Data dataKey = mapEntries.getKey(currentIndex);
+            Data value = mapEntries.getValue(currentIndex);
+            put(dataKey, value);
+            currentIndex++;
+        }
+    }
+
+    protected void put(Data dataKey, Data dataValue) {
+        MultiMapContainer container = getOrCreateContainer();
+        Collection c = (Collection) toObject(dataValue);
+        Collection<MultiMapRecord> coll = container.getOrCreateMultiMapValue(dataKey).getCollection(false);
+        Iterator it = c.iterator();
+
+        while (it.hasNext()) {
+            Object o = it.next();
+            //NB: the collection values are always passed in as unencoded objects
+            MultiMapRecord record = new MultiMapRecord(container.nextId(), isBinary() ? toData(o) : o);
```

Comment:
`o` will be already `Data`, so instead of `isBinary() ? toData(o) : o`, change this to `isBinary() ? o : toObject(o)`
(once you implement serialization as described in the comment in MultiMapProxyImpl).